control-plane/api-gateway/controllers/gatewayclass_controller.go

```diff
@@ -0,0 +1,101 @@
+package controllers
+
+import (
+	"context"
+	"time"
+
+	"github.com/go-logr/logr"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	gwv1beta1 "sigs.k8s.io/gateway-api/apis/v1beta1"
+)
+
+const (
+	GatewayClassFinalizer      = "gateway-exists-finalizer.networking.x-k8s.io"
+	GatewayClassControllerName = "hashicorp.com/consul-api-gateway-controller"
+)
+
+// GatewayClassReconciler reconciles a GatewayClass object.
+// The GatewayClass is responsible for defining the behavior of API gateways
+// which reference the given class.
+type GatewayClassReconciler struct {
+	ControllerName string
+	Log            logr.Logger
+	client.Client
+}
+
+// +kubebuilder:rbac:groups=gateway.networking.k8s.io,resources=gatewayclasses,verbs=get;list;watch;create;update;patch;delete
+// +kubebuilder:rbac:groups=gateway.networking.k8s.io,resources=gatewayclasses/status,verbs=get;update;patch
+// +kubebuilder:rbac:groups=gateway.networking.k8s.io,resources=gatewayclasses/finalizers,verbs=update
+
+// Reconcile handles the reconciliation loop for GatewayClass objects.
+func (r *GatewayClassReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	log := r.Log.WithValues("gatewayClass", req.NamespacedName)
+
+	gc := &gwv1beta1.GatewayClass{}
+
+	err := r.Client.Get(ctx, req.NamespacedName, gc)
+	if err != nil {
+		log.Error(err, "unable to get GatewayClass")
+		return ctrl.Result{}, client.IgnoreNotFound(err)
+	}
+
+	if string(gc.Spec.ControllerName) != r.ControllerName {
+		// This GatewayClass is not for this controller.
+		return ctrl.Result{}, nil
+	}
+
+	if !gc.ObjectMeta.DeletionTimestamp.IsZero() {
+		// We have a deletion request. Ensure we are not in use.
+		used, err := r.isGatewayClassInUse(ctx, gc)
+		if err != nil {
+			log.Error(err, "unable to check if GatewayClass is in use")
+			return ctrl.Result{}, err
+		}
+		if used {
+			// Requeue after 10 seconds to check again.
+			log.Info("GatewayClass is in use, cannot delete")
+			return ctrl.Result{RequeueAfter: 10 * time.Second}, nil
+		}
+		// Remove our finalizer.
+		if _, err := RemoveFinalizer(ctx, r.Client, gc, GatewayClassFinalizer); err != nil {
+			log.Error(err, "unable to remove finalizer")
+			return ctrl.Result{}, err
+		}
+		return ctrl.Result{}, nil
+	}
+
+	// We are creating or updating the GatewayClass.
+	didUpdate, err := EnsureFinalizer(ctx, r.Client, gc, GatewayClassFinalizer)
+	if err != nil {
+		log.Error(err, "unable to add finalizer")
+		return ctrl.Result{}, err
+	}
+	if didUpdate {
+		// We updated the GatewayClass, requeue to avoid another update.
+		return ctrl.Result{Requeue: true}, nil
+	}
+
```

Comment:
Here we should be validating the actual `parameterRef` object that the gatewayclass refers to. In particular:

1. any set `parametersRef` should reference only a `GatewayClassConfig` object, otherwise have a status set with `InvalidParameters` with a message saying we have the wrong parameter type
2. that `GatewayClassConfig` object should exist, otherwise also `InvalidParameters` with a message saying the parameter object doesn't exist
3. assuming that we are requiring `parametersRef` actually be set (which maybe we want to relax now that we're talking about stripping out all of the consul connection info from the config object?), then we'd also set an `InvalidParameters` condition if `parametersRef` is unspecified
4. otherwise, set `Accepted` == `True` with an `Accepted` reason.

All of this will need to be dirty-checked to make sure we don't infinitely trigger reconciliation.